pkg/typesystem/typesystem.go

```diff
@@ -1698,6 +1701,133 @@ func (t *TypeSystem) IsTuplesetRelation(objectType, relation string) (bool, erro
 	return false, nil
 }
 
+// GetEdgesFromWeightedGraph returns all edges which have a path to the source type. It's responsible for handling
+// Operator nodes, which are nodes representing Intersection (AND) or Exclusion (BUT NOT) relations. Union (OR) nodes
+// are also Operators, but we must traverse all of their edges and can't prune in advance, so this function will
+// return all relevant edges from an OR.
+// In the future we may prioritize lower weight edges in ORs, but this function is not currently doing so.
+//
+// For AND relations, we choose only the lowest weight outgoing edge, and then mark that result as "needs check".
+// E.g. If we have `rel1: a AND b AND c`, this function will return the edge with the lowest weight. If they are identical weights,
+// it will return the first edge encountered.
+//
+// For BUT NOT relations, GetEdgesFromWeightedGraph first checks if the BUT NOT applies to the source type, and if it
+// does it will mark this result as "requires check".
+// E.g. If we have `rel1: a OR b BUT NOT c` and we are searching for a "user", if 'c' does not lead to type user,
+// we do not mark as "requires check".
+// After determining whether this result will require check, GetEdgesFromWeightedGraph will prune off the last edge of the
+// Exclusion, as the right-most edge is always the BUT NOT portion, and that edge has already been accounted for.
+//
+// GetEdgesFromWeightedGraph returns a list of edges, boolean indicating whether Check is needed, and an error.
+func (t *TypeSystem) GetEdgesFromWeightedGraph(
+	targetTypeRelation string,
+	sourceType string,
+) ([]*graph.WeightedAuthorizationModelEdge, bool, error) {
+	if t.authzWeightedGraph == nil {
+		return nil, false, fmt.Errorf("weighted graph is nil")
+	}
+
+	wg := t.authzWeightedGraph
+
+	currentNode, ok := wg.GetNodeByID(targetTypeRelation)
+	if !ok {
+		return nil, false, fmt.Errorf("could not find node with label: %s", targetTypeRelation)
+	}
+
+	// This means we cannot reach the source type requested, so there are no relevant edges.
+	if !hasPathTo(currentNode, sourceType) {
+		return nil, false, nil
+	}
+
+	edges, ok := wg.GetEdgesFromNode(currentNode)
+	if !ok {
+		return nil, false, fmt.Errorf("no outgoing edges from node: %s", currentNode.GetUniqueLabel())
+	}
+
+	// needsCheck is intended to be a temporary necessity for use by list_objects/reverse_expand. There is upcoming work
+	// to remove Check from that workflow entirely, but until that's complete we need this information.
+	var needsCheck bool
+
+	if currentNode.GetNodeType() == graph.OperatorNode {
+		switch currentNode.GetLabel() {
+		case graph.ExclusionOperator: // e.g. rel1: [user, other] BUT NOT b
+			butNotEdge := edges[len(edges)-1] // this is the edge to 'b'
```

Comment:
What about:
`rel1: [user, other] but not (b or c)`

Reply:
In this case, butNotEdge will be the union node of (b or c)